src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/RenameEndpoint.cs

```diff
@@ -88,6 +100,46 @@ internal class RenameEndpoint : IRenameEndpoint
                 return null;
             }
 
+            var razorEdits = await TryGetRazorComponentRenameEditsAsync(request, requestDocumentSnapshot, codeDocument, cancellationToken).ConfigureAwait(false);
+
+            if (razorEdits is not null)
+            {
+                return razorEdits;
+            }
+
+            // If we're not doing single server rename, then we're done. C# and Html will be handled by the RenameHandler in the HtmlCSharp server.
+            if (!_languageServerFeatureOptions.SingleServerRenameSupport)
+            {
+                return null;
+            }
+
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var absoluteIndex))
+            {
+                return null;
+            }
+
+            var projection = await documentContext.GetProjectionAsync(absoluteIndex, _documentMappingService, cancellationToken).ConfigureAwait(false);
+
+            // If the language is Razor then the downstream servers won't know how to handle it anyway
+            if (projection.LanguageKind == Protocol.RazorLanguageKind.Razor)
+            {
+                return null;
+            }
+
+            var delegatedParams = new DelegatedRenameParams(
+                documentContext.Identifier,
+                projection.Position,
+                projection.LanguageKind,
+                request.NewName);
+            var delegatedRequest = await _languageServer.SendRequestAsync(LanguageServerConstants.RazorRenameEndpointName, delegatedParams).ConfigureAwait(false);
```

Comment:
Is there a pattern we could imbue across the Razor stack that encompasses:
1. Do a "Razor thing"
2. Do a "HTML or C# thing"

Simultaneously? Asking because if our own rename takes some time we're potentially delaying the computation of the C# / HTML request until its done. I played around with building aggregator models for [completion lists](https://github.com/dotnet/razor-tooling/blob/e95f3afeaca8c408744bd59e2a849fbbf63fd714/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/AggregateCompletionListProvider.cs) and the [resolving of completion items](https://github.com/dotnet/razor-tooling/blob/e95f3afeaca8c408744bd59e2a849fbbf63fd714/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/AggregateCompletionItemResolver.cs) but both of those ended up being tied to those features + required provider models which is probably a bit heavy. I'd love your thoughts on any pathways we could use to make running both computations in parallel simple, understandable and maintainable. Aka imagine @ryzngard doing the work for single server hover support. He'll probably have the same requirements

Reply:
Right now (in dual server mode) when you rename a component, we handle renaming of "SurveyPrompt" to "SurveyPrompt2", and we delegate to the Html server asking it to rename to "SurveyPrompt2". Under the debugger, by the time that request is issued, the document has already changed and the tag is already "SurveyPrompt2". Even if it wasn't though (and presumably sometimes its not, when debugging), we objectively want to handle component rename first, as we will rename the file. To me, that means the Html request is a waste of resources.

Now if we knew that our component rename was slow (do we? is it?) then I guess it makes sense (though Html currently doesn't handle rename, so the point is a little moot) so maybe I'm worrying about the wrong thing? Certainly in the case of completion where we want responses from all server, it makes sense. For rename (and hover, for example) though there is a clear heirarchy of which responses are the best, and in the case where our Razor response is best, and we can compute it without any LSP requests, that seems like a good thing.

Maybe its a question of how many functions fall into which bucket? Parallel obviously makes sense for completion, but is it the odd one out? Or is rename the odd one out? Certainly having an established pattern for common things is good.

Reply:
I think you're totally right that currently rename is fast enough for it to not be an issue. I think I was channeling the general idea of building something that's re-usable everywhere and then trying to pick a paradigm (in parallel or not execution) to operate on